spyder/plugins/editor/utils/autosave.py

```diff
@@ -106,33 +109,78 @@ def stop_autosave_timer(self):
+    def get_files_to_recover(self):
+        """
+        Get list of files to recover from pid files in autosave dir.
+
+        This returns a tuple `(files_to_recover, pid_files)`. In this tuple,
+        `files_to_recover` is a list of tuples containing the original file
+        names and the corresponding autosave file names, as recorded in the
+        pid files in the autosave directory. Any files in the autosave
+        directory which are not listed in a pid file, are also included, with
+        the original file name set to `None`. The second entry, `pid_files`,
+        is a list with the names of the pid files.
+        """
         autosave_dir = get_conf_path('autosave')
-        autosave_mapping = CONF.get('editor', 'autosave_mapping', {})
-        dialog = RecoveryDialog(autosave_dir, autosave_mapping,
-                                parent=self.editor)
+        if not os.access(autosave_dir, os.R_OK):
+            return [], []
+
+        files_to_recover = []
+        files_mentioned = []
+        pid_files = []
+        non_pid_files = []
+
+        # In Python 3, easier to use os.scandir()
+        for name in os.listdir(autosave_dir):
+            full_name = osp.join(autosave_dir, name)
+            match = re.match(r'pid([0-9]*)\.txt\Z', name)
+            if match:
+                pid_files.append(full_name)
+                logger.debug('Reading pid file: {}'.format(full_name))
+                with open(full_name) as pidfile:
+                    txt = pidfile.read()
+                    txt_as_dict = ast.literal_eval(txt)
+                    files_mentioned += [autosave for (orig, autosave)
+                                        in txt_as_dict.items()]
+                pid = int(match.group(1))
+                if is_spyder_process(pid):
+                    logger.debug('Ignoring files in {}'.format(full_name))
+                else:
+                    files_to_recover += list(txt_as_dict.items())
+            else:
+                non_pid_files.append(full_name)
+
+        # Add all files not mentioned in any pid file
+        for filename in set(non_pid_files) - set(files_mentioned):
+            files_to_recover.append((None, filename))
+            logger.debug('Added unmentioned file: {}'.format(filename))
+
+        return files_to_recover, pid_files
+
+    def try_recover_from_autosave(self):
+        """
+        Offer to recover files from autosave.
+
+        Read pid files to get a list of files that can possibly be recovered,
+        then ask the user what to do with these files, and finally remove
+        the pid files.
+        """
+        files_to_recover, pidfiles = self.get_files_to_recover()
+        dialog = RecoveryDialog(files_to_recover, parent=self.editor)
         dialog.exec_if_nonempty()
         self.recover_files_to_open = dialog.files_to_open[:]
+        for pidfile in pidfiles:
+            try:
+                os.remove(pidfile)
+            except OSError:
```

Comment:
Perhaps we should also incluce IOError here?

Reply:
Definitely. I will be glad when Python 2 disappears!